### PR TITLE
docs: add worker-host-decoupling design specification

### PR DESCRIPTION
**Pull request by jusiskin**
### What was the problem/requirement? (What/Why)

This PR builds upon #238 and #241:

>   ### Background
>   
>   `deadline-cloud-test-fixtures` provides the following class heirarchy from the abstract base class `DeadlineWorker`:
>   
>   ```txt
>   DeadlineWorker (ABC)
>   ├── EC2InstanceWorker
>   │   ├── WindowsInstanceWorkerBase
>   │   │   └── WindowsInstanceBuildWorker
>   │   └── PosixInstanceWorkerBase
>   │       └── PosixInstanceBuildWorker
>   └── DockerContainerWorker
>   ```
>   
>   These classes are responsible for managing a worker host (EC2 instance / docker container) as well as the worker agent program that is installed, configured, and ran on the host.
>   
>   The life-cycle of these are tightly-coupled. A `DeadlineWorker` instance can only be used for a single worker agent installation and configuration.
>   
>   ### Problem
>   
>   When a consuming package needs to run tests with different worker agent configurations, the tests would need to instantiate different `DeadlineWorker` instances which would then launch separate hosts. Launching EC2 hosts can take significant time.
>   
>   For example, the Deadline Cloud worker agent end-to-end tests launch have the following performance characteristics:
>   
>   | Operating System | EC2 Instance Launches | Average Launch Time | Total Launch Time |
>   | --- | --- | --- | --- |
>   | Windows | 18 | 5.4 m | 1h 37 m |
>   | Linux | 16 | 3.3 m | 53 m |
>   
>   There is a lot of time spent in these tests launching EC2 instances and preparing them. We would like to explore a solution to make these kinds of test suites more efficient.

and:

>   The feature branch (`feature_worker_host`) has been setup to manage that feature. The feature development will use a new experiment. We will be developing the feature using the [Kiro IDE](https://kiro.dev/)'s spec feature ([docs](https://kiro.dev/docs/specs/)).

A design is needed for decoupling EC2 worker host management from worker agent lifecycle management in the `deadline-cloud-test-fixtures` library. The current architecture tightly couples these concerns, preventing reuse of EC2 worker hosts with different worker agent configurations.

### What was the solution? (How)

Created a comprehensive design specification document that:

- Defines a new `WorkerHost` abstraction to separate host provisioning from agent lifecycle
- Specifies composition-based architecture where `EC2InstanceWorker` classes delegate to `WorkerHost` instances
- Details OS-specific implementations (`WindowsEC2WorkerHost`, `PosixEC2WorkerHost`)
- Establishes state management and validation for both host and agent lifecycles
- Defines 20 correctness properties to ensure proper behavior
- Includes migration strategy and testing approach

### What is the impact of this change?

This design specification will enable:
- Reusing EC2 worker hosts with different worker agent configurations
- Clear separation of concerns between host and agent management
- Independent control over host and agent lifecycles
- New pytest fixtures for host-only testing scenarios

The design identifies breaking changes that will require consumer code updates when implemented.

### How was this change tested?

This is a design document that defines the testing strategy including:
- Unit tests for interface implementations
- Property-based tests for correctness verification
- Integration tests for end-to-end scenarios
- Comprehensive test organization and framework selection

### Was this change documented?

Yes, this change adds the complete design specification including:
- Architecture diagrams and component interfaces
- State management and lifecycle validation
- Requirements traceability and correctness properties
- Migration guide for consumers
- Implementation and testing strategies

### Is this a breaking change?

No, this is only a design document. The actual implementation will introduce breaking changes as documented in the design, but this PR only adds the specification without modifying any existing code.

----

*By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.*

### CODE REVIEW

File: .kiro/specs/worker-host-decoupling/design.md [hunk 1]
```diff
@@ -0,0 +1,1487 @@
...
+    worker_id_b = worker_b.get_worker_id()
+    # Use worker_b...
+    worker_b.stop()
+```
+
+**Note**: `start()` performs complete agent startup (install, configure, start service, retrieve worker ID), and `stop()` performs complete teardown (stop service, remove state files, deleted from Deadline Cloud). This ensures the host is ready for a new agent configuration after teardown.
+
+#### worker Fixture
+
+The `worker` fixture depends on `worker_host` and provides a fully configured worker (host + agent):
```
Comment (AWS-Samuel) [inline .kiro/specs/worker-host-decoupling/design.md:1341]: I assume we'll want to scope the worker fixture to the `function` level, and the host fixture to the `session` level. Might want to include this detail

Comment (jusiskin) [reply .kiro/specs/worker-host-decoupling/design.md:1341]: Should the default be to re-install the agent for every test?

Modules/classes/tests can override the `worker` fixture to be session-scoped if they need to re-install it. This is how the worker agent e2e test suite works today.

Comment (AWS-Samuel) [reply .kiro/specs/worker-host-decoupling/design.md:1341]: We probably want to consider having a session scoped fixture and a function scoped fixture to support both use cases

Comment (jusiskin) [reply .kiro/specs/worker-host-decoupling/design.md:1341]: Usually if someone wants a function-scoped worker, they would want to modify the worker configuration between tests.

Currently, the `worker` fixture uses a session-scoped `worker_config` fixture that uses environment variables as the source of the configuration values:

https://github.com/aws-deadline/deadline-cloud-test-fixtures/blob/2f95a7579f5c7bc7bbe1368f1f499efc3126cac4/src/deadline_test_fixtures/fixtures.py#L433-L459

So we could have a function-scoped worker fixture, but it would need to depend on a function-scoped configuration that the consumer would need to provide somehow.

Comment (AWS-Samuel) [reply .kiro/specs/worker-host-decoupling/design.md:1341]: So the goal of my original comment was to point out that we need to scope the fixtures in order to get the benefit of host reuse. At a minimum we need the worker host fixture to be at a scope greater than function level in order to get the efficiency gains of keeping around the host. 

The pattern described [here](https://github.com/aws-deadline/deadline-cloud-test-fixtures/pull/245/files#diff-3305c782cf75328675e8e46b3fa3bc6b31e58c77e92ae005da32db87aee5a9d0R1196-R1207) seems to indicate we want the ability to reuse the host with multiple worker configs, but we need a mechanism to swap workers and their config (plus potentially reset some files on the host). I think we also want to expose this mechanism through fixture scopes (where basically tests that care about the worker config can request a fresh one).

If we only use session scope for the worker fixture I don't think we can reuse the host with another configuration, which goes against the `Separate control over host and agent lifecycle` requirement by effectively locking the two fixtures together.

File: .kiro/specs/worker-host-decoupling/design.md [hunk 2]
```diff
@@ -0,0 +1,1487 @@
...
+        """
+        if self._active_worker_id is not None and self._active_worker_id != worker_id:
+            raise RuntimeError(
+                f"Cannot start worker agent: another worker (id={self._active_worker_id}) "
+                f"already has an agent running on this host. "
+                f"Call stop() on that worker first."
+            )
+        self._active_worker_id = worker_id
+    
+    def _release_from_worker(self, worker_id: int) -> None:
```
Comment (lucaseck) [inline .kiro/specs/worker-host-decoupling/design.md:183]: I forget if our tests on these hosts run serially or not, but is the assumption then that each individual test will have to fully wait for the host to be released and agent shutdown before the next test can be started? Assuming that the tests need a different worker agent configuration. If they are sequential do we already try and say run in an order to maximize use of a matching worker configuration

File: .kiro/specs/worker-host-decoupling/design.md [hunk 3]
```diff
@@ -0,0 +1,1487 @@
...
+### Unit Testing
+
+Unit tests will verify specific behaviors and edge cases:
+
+1. **WorkerHost Interface Tests**: Verify each EC2 WorkerHost implementation (Windows, POSIX) correctly implements the interface
+2. **DeadlineWorker Delegation Tests**: Verify EC2InstanceWorker classes correctly delegate to WorkerHost
+3. **Configuration Application Tests**: Verify worker agent configurations are correctly applied
+4. **Error Handling Tests**: Verify appropriate exceptions are raised with correct diagnostic information
+5. **Operating System Validation Tests**: Verify operating system compatibility checks work correctly
+6. **Worker Host Ownership Tests**: Verify host claiming/releasing behavior and ownership validation
```
Comment (lucaseck) [inline .kiro/specs/worker-host-decoupling/design.md:991]: To my earlier comment I think our tests are sequential right now, but if we are running in parallel I assume there will be some sort of locking etc so that we don't end up with multiple running on the same instance?

File: .kiro/specs/worker-host-decoupling/design.md (hunk 2 above)
Comment (jusiskin) [reply .kiro/specs/worker-host-decoupling/design.md:183]: >   I forget if our tests on these hosts run serially or not

I assume you are referring to the worker agent e2e tests, right? We do not use pytest-xdist for those because:

*   each pytest worker runs in a separate pytest session and they have independent fixtures that are not re-used ([ref](https://pytest-xdist.readthedocs.io/en/stable/how-to.html#making-session-scoped-fixtures-execute-only-once))

    >   This means that tests in different processes requesting a high-level scoped fixture (for example `session`) will execute the fixture code more than once, which breaks expectations and might be undesired in certain situations.

*   the tests use persistent queues and fleets that are shared by the tests. The tests assert various properties on the worker host after running jobs, but Deadline Cloud will schedule jobs to any worker in a fleet associated with the job's queue. If there are multiple workers, we would need isolated queues and QFAs for each pytest worker's workers to ensure that our jobs ran on the correct worker.

My understanding is that pytest group tests that share the same fixtures together to minimize active fixtures. The [pytest documentation](https://docs.pytest.org/en/6.2.x/fixture.html#automatic-grouping-of-tests-by-fixture-instances) on the subject does say that:

>   **pytest minimizes the number of active fixtures during test runs.** If you have a parametrized fixture, then all the tests using it will first execute with one instance and then **finalizers are called before the next fixture instance is created**. ...

This will require some prototyping/experimentation. But if this doesn't work, we have a `pytest_collection_modifyitems` hook in the worker agent that we used for similar purposes:

https://github.com/aws-deadline/deadline-cloud-worker-agent/blob/c5b1c2d128d0015badf5d060bc7fed9f74920eee/test/e2e/conftest.py#L550-L580

We could document this approach for consumers of this library or maybe the library could implement a general-purpose hook for consumers using fixtures for host re-use.

File: .kiro/specs/worker-host-decoupling/design.md (hunk 3 above)
Comment (jusiskin) [reply .kiro/specs/worker-host-decoupling/design.md:991]: Yes, this is the intention of the [worker host ownership tracking](https://github.com/jusiskin/deadline-cloud-test-fixtures/blob/3d3a19cf40607daab232b7b7909955c78eb22e22/.kiro/specs/worker-host-decoupling/design.md#worker-host-ownership-tracking) section.

Are you saying we should have guards when tests are running in parallel? As mentioned in the other comment, if consumers are using `pytest-xdist` then the parallel pytest workers run in their own independent pytest session and fixtures are not shared (including worker hosts).

File: .kiro/specs/worker-host-decoupling/design.md (hunk 1 above)
Comment (jusiskin) [reply .kiro/specs/worker-host-decoupling/design.md:1341]: You are correct. If someone wants to re-use the host using fixtures, then their worker fixture needs to have more fine-grained scope than the worker host.

### Valid Host Re-Use Scopes

| Host Scope | Worker Scope |
| --- | --- |
| `session` | `module` |
| `session` | `class` |
| `session` | `function` |
| `module` | `class` |
| `module` | `function` |
| `class` | `function` |

### Non-valid Host Re-Use Scopes

| Host Scope | Worker Scope |
| --- | --- |
| `session` | `session` |
| `module` | `module` |
| `class` | `class` |
| `function` | `function` |

The reason I'm pushing back against changing the scope of the `worker` fixture provided by the `deadline-cloud-test-fixtures` library is:

1.  **It would make the default behavior slower** &mdash; every test using the `worker` fixture would by default have the worker agent setup and uninstalled between each test. This would slow down tests that do not require a special worker configuration.
2.  **We would need to design how the worker is configured per-function** &mdash; we currently use env vars to configure the worker agent (see my earlier comment in this thread). But if the default scope for the worker agent is `function`, then we either need consumers to override the `worker_config` fixture with a function-scoped fixture, or provide a some way to configure this configuration from inputs outside of pytest fixtures (like we used env vars)

File: .kiro/specs/worker-host-decoupling/design.md (hunk 3 above)
Comment (lucaseck) [reply .kiro/specs/worker-host-decoupling/design.md:991]: Makes sense, I did see that we had the ownership tracking I was just curious if there could be any race conditions there that we would need to handle

File: .kiro/specs/worker-host-decoupling/design.md (hunk 2 above)
Comment (lucaseck) [reply .kiro/specs/worker-host-decoupling/design.md:183]: Cool that answers my question then and our tests at least should already be being grouped together to maximize reuse

**Review by lucaseck — APPROVED** [review]

**Review by AWS-Samuel — APPROVED** [review]
